File: server/voxelizer.py
"""
voxelizer.py — 3-D voxelization and volumetric diff (server-side)

Mirrors lib/voxelizer.js exactly, with numpy for vectorised operations.

ALGORITHM (identical to the JS version):
  1. make_grid_def(points, vox_size)
     Compute lon/lat/h step sizes from the average latitude of all points.

  2. build_surface(points, grid_def, polygon)
     Bin each point into its (iLon, iLat, iH) voxel cell.
     Returns a set of voxel keys for the sparse surface shell.

  3. solidify(surface_a, surface_b, grid_def)
     For each (iLon, iLat) column, fill from per-column floor up to maxH.
     Per-column floor = min(minH_A, minH_B) — prevents a single noisy
     outlier in one column from dragging the floor down everywhere else.

  4. diff_solid(solid_a, solid_b)
     Keys in B but not A → 'added'
     Keys in A but not B → 'removed'
     Returns voxel coordinate dicts {iLon, iLat, iH, type}.
"""

from __future__ import annotations

from typing import Optional

import numpy as np


# ── Type alias ────────────────────────────────────────────────────────────
# A surface/solid is a dict mapping "iLon,iLat,iH" → (iLon, iLat, iH) tuple.
VoxelMap = dict[str, tuple[int, int, int]]


# ── Grid definition ───────────────────────────────────────────────────────

def make_grid_def(points: np.ndarray, vox_size: float) -> dict:
    """
    Compute the shared voxel grid definition.

    Parameters
    ----------
    points   : (N, 3) float64  [lon°, lat°, h m]
    vox_size : cell size in metres

    Returns
    -------
    dict with keys: lon_step, lat_step, h_step  (all in degrees or metres)
    """
    if points.shape[0] == 0:
        step = vox_size / 111_000
        return {"lon_step": step, "lat_step": step, "h_step": vox_size}

    valid = points[np.isfinite(points).all(axis=1)]
    if valid.shape[0] == 0:
        step = vox_size / 111_000
        return {"lon_step": step, "lat_step": step, "h_step": vox_size}

    avg_lat = float(valid[:, 1].mean())
    cos_lat = np.cos(np.radians(avg_lat))

    return {
        "lon_step": vox_size / (111_000 * cos_lat),   # degrees lon per voxel
        "lat_step": vox_size / 111_000,               # degrees lat per voxel
        "h_step":   vox_size,                         # metres per voxel
    }


# ── Step 1: surface shell ─────────────────────────────────────────────────

def _pip(lon: np.ndarray, lat: np.ndarray, poly: list[dict]) -> np.ndarray:
    """
    Vectorised point-in-polygon (ray casting) for arrays of lon/lat.

    Parameters
    ----------
    lon, lat : 1-D float64 arrays of length N
    poly     : list of {lon, lat} dicts (closed polygon)

    Returns
    -------
    Boolean mask of length N — True = inside polygon
    """
    inside = np.zeros(len(lon), dtype=bool)
    n = len(poly)
    j = n - 1
    for i in range(n):
        xi, yi = poly[i]["lon"], poly[i]["lat"]
        xj, yj = poly[j]["lon"], poly[j]["lat"]
        # Ray-casting condition
        cond = ((yi > lat) != (yj > lat)) & (
            lon < (xj - xi) * (lat - yi) / (yj - yi) + xi
        )
        inside ^= cond
        j = i
    return inside


def build_surface(
    points:   np.ndarray,
    grid_def: dict,
    polygon:  Optional[list[dict]] = None,
) -> VoxelMap:
    """
    Bin raw geodetic points into a sparse voxel occupancy map.

    Parameters
    ----------
    points   : (N, 3) float64  [lon°, lat°, h m]
    grid_def : output of make_grid_def()
    polygon  : optional list of {lon, lat} dicts; None = no filter

    Returns
    -------
    dict  key="iLon,iLat,iH"  value=(iLon, iLat, iH)
    """
    lon_step = grid_def["lon_step"]
    lat_step = grid_def["lat_step"]
    h_step   = grid_def["h_step"]

    # Drop NaN / Inf
    mask = np.isfinite(points).all(axis=1)
    pts  = points[mask]

    if pts.shape[0] == 0:
        return {}

    lon = pts[:, 0]
    lat = pts[:, 1]
    h   = pts[:, 2]

    # Apply polygon filter
    if polygon and len(polygon) >= 3:
        in_poly = _pip(lon, lat, polygon)
        lon, lat, h = lon[in_poly], lat[in_poly], h[in_poly]

    if len(lon) == 0:
        return {}

    i_lon = np.floor(lon / lon_step).astype(np.int64)
    i_lat = np.floor(lat / lat_step).astype(np.int64)
    i_h   = np.floor(h   / h_step  ).astype(np.int64)

    surface: VoxelMap = {}
    for il, ia, ih in zip(i_lon, i_lat, i_h):
        key = f"{il},{ia},{ih}"
        if key not in surface:
            surface[key] = (int(il), int(ia), int(ih))

    return surface


# ── Step 2: solidification ────────────────────────────────────────────────

def solidify(
    surface_a: VoxelMap,
    surface_b: VoxelMap,
    grid_def:  dict,          # kept for API symmetry with JS; not used here
) -> tuple[VoxelMap, VoxelMap]:
    """
    Fill each (iLon, iLat) column from per-column floor up to maxH.

    Returns
    -------
    (solid_a, solid_b) — two VoxelMaps of filled-solid voxels
    """
    # ── Column stats per dataset ──────────────────────────────────────────
    def _col_stats(surface: VoxelMap) -> dict[str, dict]:
        cols: dict[str, dict] = {}
        for (il, ia, ih) in surface.values():
            ck = f"{il},{ia}"
            if ck not in cols:
                cols[ck] = {"il": il, "ia": ia, "min_h": ih, "max_h": ih}
            else:
                if ih < cols[ck]["min_h"]: cols[ck]["min_h"] = ih
                if ih > cols[ck]["max_h"]: cols[ck]["max_h"] = ih
        return cols

    cols_a = _col_stats(surface_a)
    cols_b = _col_stats(surface_b)

    # ── Per-column floor = min(minH_A, minH_B) ───────────────────────────
    col_floor: dict[str, int] = {}
    for ck, col in cols_a.items():
        col_floor[ck] = col["min_h"]
    for ck, col in cols_b.items():
        prev = col_floor.get(ck)
        col_floor[ck] = col["min_h"] if prev is None else min(prev, col["min_h"])

    # ── Fill columns floor → maxH ─────────────────────────────────────────
    def _fill(cols: dict[str, dict]) -> VoxelMap:
        solid: VoxelMap = {}
        for ck, col in cols.items():
            il, ia  = col["il"], col["ia"]
            floor_h = col_floor.get(ck, 0)
            for ih in range(floor_h, col["max_h"] + 1):
                key = f"{il},{ia},{ih}"
                if key not in solid:
                    solid[key] = (il, ia, ih)
        return solid

    solid_a = _fill(cols_a)
    solid_b = _fill(cols_b)

    print(
        f"[voxelizer] Solid voxels — A: {len(solid_a)}, B: {len(solid_b)}"
    )
    return solid_a, solid_b


# ── Step 3: diff ──────────────────────────────────────────────────────────

def diff_solid(
    solid_a: VoxelMap,
    solid_b: VoxelMap,
) -> tuple[list[dict], list[dict]]:
    """
    Compute the volumetric diff between two solid voxel maps.

    Returns
    -------
    (added, removed)
    Each element is a list of dicts: { iLon, iLat, iH, type }
    """
    added:   list[dict] = []
    removed: list[dict] = []

    for key, (il, ia, ih) in solid_b.items():
        if key not in solid_a:
            added.append({"iLon": il, "iLat": ia, "iH": ih, "type": "added"})

    for key, (il, ia, ih) in solid_a.items():
        if key not in solid_b:
            removed.append({"iLon": il, "iLat": ia, "iH": ih, "type": "removed"})

    return added, removed

File: frontend/src/lib/voxelizer.js
/**
 * lib/voxelizer.js
 *
 * Pure 3-D voxelization and volumetric diff.
 * No Cesium, no React, no browser APIs — safe to run in a Web Worker or Node.js.
 *
 * The only external dep is polygonUtils.js (pip).
 *
 * ── ALGORITHM OVERVIEW ────────────────────────────────────────────────────
 *
 *  1. makeGridDef(points, voxelSize)
 *     Compute a shared geographic grid definition from all input points.
 *     lonStep / latStep convert metres → degrees, corrected for latitude.
 *
 *  2. buildSurface(points, gridDef, polygon)
 *     Bin each point into its voxel cell (iLon, iLat, iH).
 *     Returns a Map<key, {iLon, iLat, iH}> — the sparse surface shell.
 *
 *  3. solidify(surfaceA, surfaceB, gridDef)
 *     For each (iLon, iLat) column, fill every voxel from the
 *     per-column floor up to its dataset's max height.
 *
 *     WHY PER-COLUMN FLOOR:
 *     A global floor (= min iH across all points) breaks if any tile has
 *     a single noisy outlier below the terrain — it drags the fill floor
 *     down for every column everywhere, producing underground boxes.
 *     Per-column floor = min iH that column has in *either* dataset, so
 *     one column's outlier cannot affect any other column.
 *
 *  4. diffSolid(solidA, solidB)
 *     Keys in B but not A → added.
 *     Keys in A but not B → removed.
 *     Returns { added: Cartesian3[], removed: Cartesian3[] } — voxel centres
 *     in ECEF, ready to be passed to renderVoxelDiff().
 */

import { pip } from './polygonUtils'

// ── Grid definition ────────────────────────────────────────────────────────

/**
 * Compute the shared voxel grid definition from all input points.
 *
 * @param {Array<{lon,lat,h}>} points   — combined raw points from both dates
 * @param {number}             voxSize  — cell size in metres
 * @returns {{ lonStep, latStep, hStep }}
 */
export function makeGridDef(points, voxSize) {
  const valid = points.filter(p => isFinite(p.lon) && isFinite(p.lat) && isFinite(p.h))
  if (!valid.length) return { lonStep: voxSize / 111000, latStep: voxSize / 111000, hStep: voxSize }

  const avgLat = valid.reduce((s, p) => s + p.lat, 0) / valid.length
  const cosLat = Math.cos(avgLat * Math.PI / 180)

  return {
    lonStep: voxSize / (111000 * cosLat),   // degrees lon per voxel
    latStep: voxSize / 111000,              // degrees lat per voxel
    hStep:   voxSize,                       // metres per voxel
  }
}

// ── Step 1: surface shell ──────────────────────────────────────────────────

/**
 * Bin raw geodetic points into a voxel occupancy map (the surface shell).
 * Points outside the polygon are discarded.
 *
 * @param {Array<{lon,lat,h}>}              points
 * @param {{ lonStep, latStep, hStep }}     gridDef
 * @param {Array<{lon,lat}>|null}           polygon  — null = no filter
 * @returns {Map<string, {iLon, iLat, iH}>}
 */
export function buildSurface(points, gridDef, polygon) {
  const { lonStep, latStep, hStep } = gridDef
  const map = new Map()

  for (const { lon, lat, h } of points) {
    if (!isFinite(lon) || !isFinite(lat) || !isFinite(h)) continue
    if (polygon && !pip(lon, lat, polygon)) continue

    const iLon = Math.floor(lon / lonStep)
    const iLat = Math.floor(lat / latStep)
    const iH   = Math.floor(h   / hStep)
    const key  = `${iLon},${iLat},${iH}`
    if (!map.has(key)) map.set(key, { iLon, iLat, iH })
  }

  return map
}

// ── Step 2: solidification ─────────────────────────────────────────────────

/**
 * Convert two sparse surface maps into solid volume maps by filling each
 * (iLon, iLat) column from its per-column floor up to its max height.
 *
 * The returned Maps have Cesium.Cartesian3 values (voxel centres in ECEF).
 * Requires window.Cesium to be set (cesiumInit.js sets it before any diff runs).
 *
 * @param {Map} surfaceA
 * @param {Map} surfaceB
 * @param {{ lonStep, latStep, hStep }} gridDef
 * @returns {{ solidA: Map<string, Cartesian3>, solidB: Map<string, Cartesian3> }}
 */
export function solidify(surfaceA, surfaceB, gridDef) {
  const { lonStep, latStep, hStep } = gridDef

  // ── a) Build per-column min/max for each dataset ──────────────────────
  function buildColStats(surface) {
    const cols = new Map()   // colKey → { iLon, iLat, minH, maxH }
    for (const { iLon, iLat, iH } of surface.values()) {
      const colKey = `${iLon},${iLat}`
      if (!cols.has(colKey)) {
        cols.set(colKey, { iLon, iLat, minH: iH, maxH: iH })
      } else {
        const col = cols.get(colKey)
        if (iH < col.minH) col.minH = iH
        if (iH > col.maxH) col.maxH = iH
      }
    }
    return cols
  }

  const colsA = buildColStats(surfaceA)
  const colsB = buildColStats(surfaceB)

  // ── b) Per-column floor = min(minH_A, minH_B) ────────────────────────
  const colFloor = new Map()   // colKey → lowest iH in either dataset

  for (const [colKey, col] of colsA) colFloor.set(colKey, col.minH)
  for (const [colKey, col] of colsB) {
    const prev = colFloor.get(colKey)
    colFloor.set(colKey, prev !== undefined ? Math.min(prev, col.minH) : col.minH)
  }

  // ── c) Fill each column floor→maxH, compute correct voxel centre ──────
  function fillCols(cols) {
    const solid = new Map()
    for (const [colKey, { iLon, iLat, maxH }] of cols) {
      const floorH = colFloor.get(colKey) ?? 0
      for (let iH = floorH; iH <= maxH; iH++) {
        const key = `${iLon},${iLat},${iH}`
        if (!solid.has(key)) {
            solid.set(key, { iLon, iLat, iH })
        }
      }
    }
    return solid
  }

  const solidA = fillCols(colsA)
  const solidB = fillCols(colsB)

  console.log(`[voxelizer] Solid voxels — A: ${solidA.size}, B: ${solidB.size}`)
  return { solidA, solidB }
}

// ── Step 3: diff ───────────────────────────────────────────────────────────

/**
 * Compute volumetric diff between two solid voxel maps.
 *
 * @param {Map<string, { iLon, iLat, iH }>} solidA
 * @param {Map<string, { iLon, iLat, iH }>} solidB
 * @returns {{ added: Array<{center, type}>, removed: Array<{center, type}> }}
 */
export function diffSolid(solidA, solidB) {
  const added   = []
  const removed = []

  for (const [key, voxel] of solidB) {
    if (!solidA.has(key)) added.push({ voxel, type: 'added' })
  }
  for (const [key, voxel] of solidA) {
    if (!solidB.has(key)) removed.push({ voxel, type: 'removed' })
  }

  return { added, removed }
}